TailLight/SetTaillightBlack.cpp

```diff
@@ -0,0 +1,250 @@
+#include "driver.h"
+#include <Hidport.h>
+
+#include "debug.h"
+#include "SetBlack.h"
+
+EVT_WDF_REQUEST_COMPLETION_ROUTINE  SetBlackCompletionRoutine;
+EVT_WDF_WORKITEM                    SetBlackWorkItem;
+
+NTSTATUS CreateWorkItemForIoTargetOpenDevice(WDFDEVICE device)
+    /*++
+
+    Routine Description:
+
+        Creates a WDF workitem to do the SetBlack() call after the driver
+        stack has initialized.
+
+    Arguments:
+
+        Device - Handle to a pre-allocated WDF work item.
+
+    Requirements:
+        Must be synchronized to the device.
+
+    --*/
+{
+
+    //TRACE_FN_ENTRY
+    
+    WDFWORKITEM hWorkItem = 0;
+    NTSTATUS status = STATUS_PNP_DRIVER_CONFIGURATION_INCOMPLETE;
+    {
+        WDF_WORKITEM_CONFIG        workItemConfig;
+        WDF_OBJECT_ATTRIBUTES      workItemAttributes;
+        WDF_OBJECT_ATTRIBUTES_INIT(&workItemAttributes);
+        workItemAttributes.ParentObject = device;
+
+        DEVICE_CONTEXT* pDeviceContext = WdfObjectGet_DEVICE_CONTEXT(device);
+
+        // It's possible to get called twice. Been there, done that?
+        if (pDeviceContext->fSetBlackSuccess) {
+            return STATUS_SUCCESS;
+        }
+
+        WDF_WORKITEM_CONFIG_INIT(&workItemConfig, SetBlackWorkItem);
+
+        status = WdfWorkItemCreate(&workItemConfig,
+            &workItemAttributes,
+            &hWorkItem);
+
+        if (!NT_SUCCESS(status)) {
+            KdPrintEx((DPFLTR_IHVDRIVER_ID,
+                DPFLTR_ERROR_LEVEL,
+                "TailLight: Workitem creation failure 0x%x\n",
+                status));
+            return status; // Maybe better luck next time.
+        }
+
+        InterlockedIncrement((PLONG)(&pDeviceContext->fSetBlackSuccess));
+    }
+
+    WdfWorkItemEnqueue(hWorkItem);
+
+    //TRACE_FN_EXIT
+
+    return status;
+}
+
+static NTSTATUS TryToOpenIoTarget(WDFIOTARGET target, 
+    DEVICE_CONTEXT& DeviceContext) {
+
+    PAGED_CODE();
+
+    WDF_IO_TARGET_OPEN_PARAMS   openParams = {};
+    WDF_IO_TARGET_OPEN_PARAMS_INIT_CREATE_BY_NAME(
+        &openParams,
+        &DeviceContext.PdoName,
+        FILE_WRITE_ACCESS);
+
+    openParams.ShareAccess = FILE_SHARE_WRITE | FILE_SHARE_READ;
+
+    NTSTATUS status = STATUS_UNSUCCESSFUL;
+
+    // Ensure freed if fails.
+    status = WdfIoTargetOpen(target, &openParams);
+    if (!NT_SUCCESS(status)) {
+        KdPrintEx((DPFLTR_IHVDRIVER_ID,
+            DPFLTR_ERROR_LEVEL,
+            "TailLight: 0x%x while opening the I/O target from worker thread\n",
+            status));
+        NukeWdfHandle<WDFIOTARGET>(target);
```

Comment:
I don't think you should delete `target` here, since the object is passed as an input argument to the function. Please instead move the `NukeWdfHandle` call to the client code, and call it if `TryToOpenIoTarget` fails.

Reply:
Done. Now in SetBlackAsync(). I'm taking more breaks to prevent such errors.

Reply:
What do you mean by "done"? The pull request have not yet been updated to address the comment.